lib/keila/templates/html_text_extractor.ex

```diff
@@ -0,0 +1,140 @@
+defmodule Keila.Templates.HtmlTextExtractor do
+  @doc """
+  Processes an HTML fragment tree and returns it as a string.
+  """
+  @spec extract_text(Keila.Templates.Html.t(), String.t()) :: String.t()
+  def extract_text(html, selector \\ "body") do
+    Floki.find(html, selector)
+    |> do_extract_text(0)
+    |> concat_extracted_text()
+    |> then(fn string -> string <> "\n" end)
+  end
+
+  defp do_extract_text(html, list_level) when is_list(html) do
+    Enum.map(html, &do_extract_text(&1, list_level))
+  end
+
+  defp do_extract_text({"br", _, _}, _), do: ["\n"]
+
+  defp do_extract_text({"ul", _, content}, list_level) do
+    content
+    |> Enum.filter(&match?({"li", _, _}, &1))
+    |> Enum.map(fn {"li", _attrs, content} ->
+      indentation = String.pad_trailing("", (list_level + 1) * 2, " ")
+
+      text_content =
+        content
+        |> do_extract_text(list_level + 1)
+        |> concat_extracted_text()
+        |> String.replace(~r{\n(\w)}, "\n#{indentation}\\g{1}")
+
+      ["- ", text_content, "\n"]
+    end)
+    |> then(fn content -> ["\n" | content] end)
+  end
+
+  defp do_extract_text({"ol", _attrs, content}, list_level) do
+    content
+    |> Enum.filter(&match?({"li", _, _}, &1))
+    |> Enum.with_index()
+    |> Enum.map(fn {{"li", _attrs, content}, i} ->
+      indentation = String.pad_trailing("", (list_level + 1) * 2, " ")
+
+      text_content =
+        content
+        |> do_extract_text(list_level + 1)
+        |> concat_extracted_text()
+        |> String.replace(~r{\n(\w)}, "\n#{indentation}\\g{1}")
+
+      ["#{i + 1}. ", text_content, "\n"]
+    end)
+    |> then(fn content -> ["\n" | content] end)
+  end
+
+  defp do_extract_text({"h" <> level, _, content}, list_level)
+       when level in ["1", "2", "3", "4", "5"] do
+    padding =
+      level |> String.to_integer() |> then(fn n -> String.pad_trailing("", n, "#") <> " " end)
+
+    [padding | do_extract_text(content, list_level) ++ ["\n"]]
+  end
+
+  defp do_extract_text({"hr", _attrs, _content}, _list_level) do
+    "\n---\n\n"
+  end
+
+  defp do_extract_text({"a", attrs, content}, list_level) do
+    href = html_attribute(attrs, "href")
+    title = html_attribute(attrs, "title")
+    content = do_extract_text(content, list_level)
+
+    case title do
```

Comment:
This case could be clearer with this:

```
if title do
    ["[", content, "]", ~s{(#{href} "#{title}")}]
  else
    ["[", content, "]", "(#{href})"]
  end
```

Reply:
I think it’s mostly a matter of preference here^^ Erlang folks would probably prefer the `case`, people from other backgrounds the `if`.